### PR TITLE
Add swap widget

### LINKED ISSUE #84: [Assets swap] Transform to widget and put on My assets page
On "My assets" page we have the same swap widget as on the "Assets Swap" page. The only difference here is that once we change the tokens to swap - we shouldn't update the URL params

Image

### PR DESCRIPTION
**Pull request by VadimSaveljev**

### CODE REVIEW

File: src/components/Pages/MyAssets/Pools/CreatePool.tsx [hunk 1]
```diff
@@ -37,10 +37,13 @@ const CreatePool = () => {
     async (event: FormEvent) => {
       event.preventDefault();
       const selectedPoolToken = availablePoolTokens?.[Number(newPoolToken)];
-      if (api && selectedPoolToken) {
+      if (api && selectedPoolToken && nativeBalance) {
         // validate user has enough funds to cover the setup fee
         const poolSetupFee = api.consts.assetConversion?.poolSetupFee ?? null;
         const existentialDeposit = api.consts.balances.existentialDeposit;
+
+        // TODO check later if we can add a proper type for poolSetupFee
```
Comment (jsidorenko) [inline src/components/Pages/MyAssets/Pools/CreatePool.tsx:45]: What if we put `const poolSetupFee: BN | null`?

File: src/components/Pages/Swap/Swap.tsx [hunk 2]
```diff
@@ -84,19 +89,103 @@ const Swap = ({
   const [formDisabled, setFormDisabled] = useState<boolean>(true);
   const [swapType, setSwapType] = useState<SwapTypes | null>(null);
 
-  const reserves = useMemo(
+  const reserves: [BN, BN] = useMemo(
```
Comment (jsidorenko) [inline src/components/Pages/Swap/Swap.tsx]:
```suggestion
  const reserves: PoolReserves = useMemo(
```

File: src/components/Pages/Swap/Swap.tsx [hunk 3]
```diff
@@ -266,13 +303,31 @@ const Swap = ({
...
       <ModalStatus />
+
       <Form onSubmit={submitSwap}>
         <Form.Group className='mb-3'>
-          <Form.Label>{asset1Metadata.symbol.toUpperCase()}:</Form.Label>
+          <Form.Select
+            className='mb-3'
+            value={multiAssetToParam(asset1)}
+            onChange={(event) =>
+              handleTokenChange(toMultiAsset(api.createType('AssetId', event.target.value), api), asset2)
```
Comment (jsidorenko) [inline src/components/Pages/Swap/Swap.tsx]:
```suggestion
              handleTokenChange(parseAssetParam(event.target.value, api), asset2)
```

File: src/hooks/useAssets.ts [hunk 4]
```diff
@@ -443,17 +482,21 @@ export const useAssets = () => {
         supply: (await api.query.balances?.totalIssuance?.().then((r) => r.toBn())) ?? null,
       });
 
-      const [details, metadata]: [TokensDetailsMap, TokenMetadata[]] = await Promise.all([
+      const [details, metadata]: [TokensDetailsMap | undefined, TokenMetadata[]] = await Promise.all([
         fetchAllTokensDetails(),
         fetchAllTokensMetadata(),
       ]);
 
       result = [
```
Comment (jsidorenko) [inline src/hooks/useAssets.ts]: You can simply wrap that `result = [....]` with `if (details) {`

File: src/hooks/useAssets.ts [hunk 5]
```diff
@@ -443,17 +482,21 @@ export const useAssets = () => {
...
         ...result,
-        ...metadata.map((token) => ({
-          ...token,
-          supply: details?.get(token.id.asAsset.toNumber()) ?? null,
-        })),
+        ...metadata.map((token) => {
+          let supply = null;
+          if (details) {
```
Comment (jsidorenko) [inline src/hooks/useAssets.ts]: and remove that wrap from here

File: src/components/Pages/Swap/Swap.tsx [hunk 6]
```diff
@@ -124,10 +213,10 @@ const Swap = ({
...
 
       // validate user has enough funds
-      if (!formError && asset1Balance.lt(amount1)) {
+      if (!formError && asset1Balance?.lt(amount1)) {
         formError = StatusMessages.POOL_AMOUNT_TOO_HIGH;
       }
-      if (!formError && asset2Balance.lt(amount2)) {
+      if (!formError && asset2Balance?.lt(amount2)) {
         formError = StatusMessages.POOL_AMOUNT_TOO_HIGH;
       }
```
Comment (jsidorenko) [inline src/components/Pages/Swap/Swap.tsx]:
```suggestion
```

File: src/components/Pages/MyAssets/Pools/CreatePool.tsx (hunk 1 above)
Comment (VadimSaveljev) [reply src/components/Pages/MyAssets/Pools/CreatePool.tsx:45]: tried that
Type 'Codec' is missing the following properties from type 'BN': clone, toNumber, toArray, toArrayLike, and 82 more.

Comment (jsidorenko) [reply src/components/Pages/MyAssets/Pools/CreatePool.tsx:45]: Ok, that means we need to wait for the proper pjs support of the assets conversion pallet

**Review by jsidorenko — APPROVED** [review]

**Comment by jsidorenko** [comment]
WD

**Comment by VadimSaveljev** [comment]
> WD

WD-40-Multi-Use-Product-Gif

**Pull request merged**